polkadot/parachain/src/wasm.rs

```diff
@@ -0,0 +1,166 @@
+	}
+	errors {
+		/// Call data too big. WASM32 only has a 32-bit address space.
+		ParamsTooLarge(len: usize) {
+			description("Validation parameters took up too much space to execute in WASM"),
+			display("Validation parameters took up {} bytes, max allowed by WASM is {}", len, i32::max_value()),
+		}
+		/// Bad return data or type.
+		BadReturn {
+			description("Validation function returned invalid data."),
+			display("Validation function returned invalid data."),
+		}
+	}
+}
+
+struct Resolver {
+	max_memory: u32, // in pages.
+	memory: RefCell<Option<MemoryRef>>,
+}
+
+impl ModuleImportResolver for Resolver {
+	fn resolve_memory(
+		&self,
+		field_name: &str,
+		descriptor: &MemoryDescriptor,
+	) -> Result<MemoryRef, WasmError> {
+		if field_name == "memory" {
+			let effective_max = descriptor.maximum().unwrap_or(self.max_memory);
+			if descriptor.initial() > self.max_memory || effective_max > self.max_memory {
+				Err(WasmError::Instantiation("Module requested too much memory".to_owned()))
+			} else {
+				let mem = MemoryInstance::alloc(
+					memory_units::Pages(descriptor.initial() as usize),
+					descriptor.maximum().map(|x| memory_units::Pages(x as usize)),
+				)?;
+				*self.memory.borrow_mut() = Some(mem.clone());
+				Ok(mem)
+			}
+		} else {
+			Err(WasmError::Instantiation("Memory imported under unknown name".to_owned()))
+		}
+	}
+}
+
+/// Validate a candidate under the given validation code.
+///
+/// This will fail if the validation code is not a proper parachain validation module.
+pub fn validate_candidate(validation_code: &[u8], params: ValidationParams) -> Result<ValidationResult, Error> {
+	use wasmi::LINEAR_MEMORY_PAGE_SIZE;
+
+	// maximum memory in bytes
+	const MAX_MEM: u32 = 1024 * 1024 * 1024; // 1 GiB
+
+	// instantiate the module.
+	let (module, memory) = {
+		let module = Module::from_buffer(validation_code)?;
+
+		let module_resolver = Resolver {
+			max_memory: MAX_MEM / LINEAR_MEMORY_PAGE_SIZE.0 as u32,
+			memory: RefCell::new(None),
+		};
+
+		let module = ModuleInstance::new(
+			&module,
+			&wasmi::ImportsBuilder::new().with_resolver("env", &module_resolver),
+		)?.run_start(&mut wasmi::NopExternals).map_err(WasmError::Trap)?;
+
+		let memory = module_resolver.memory.borrow_mut()
+			.as_ref()
+			.ok_or_else(|| WasmError::Instantiation("No imported memory instance".to_owned()))?
+			.clone();
+
+		(module, memory)
+	};
+
+	// allocate call data in memory.
+	let (offset, len) = {
+		let encoded_call_data = params.encode();
+
+		// hard limit from WASM.
+		if encoded_call_data.len() > i32::max_value() as usize {
+			bail!(ErrorKind::ParamsTooLarge(encoded_call_data.len()));
+		}
+
+		let call_data_pages = (encoded_call_data.len() / LINEAR_MEMORY_PAGE_SIZE.0) +
+			(encoded_call_data.len() % LINEAR_MEMORY_PAGE_SIZE.0);
+
+		let call_data_pages = wasmi::memory_units::Pages(call_data_pages);
+
+		if memory.current_size() < call_data_pages {
+			memory.grow(call_data_pages - memory.current_size())?;
```

Comment:
Shouldn't that grow in some larger chunks to prevent frequent re-allocs?

Reply:
the call data is only passed once at the entry point of the parachain valcode. what reallocs do you mean?